DependencyInjection/Compiler/DiscoveryPass.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Http\HttplugBundle\DependencyInjection\Compiler;
+
+use Http\Client\HttpClient;
+use Http\HttplugBundle\Util\HttplugFactory;
+use Http\Message\MessageFactory;
+use Http\Message\StreamFactory;
+use Http\Message\UriFactory;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Adds fallback and discovery services.
+ *
+ * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
+ */
+final class DiscoveryPass implements CompilerPassInterface
+{
+    /**
+     * Fallback services and classes.
+     *
+     * @var array
+     */
+    private $services = [
+        'client' => HttpClient::class,
+        'message_factory' => MessageFactory::class,
+        'uri_factory' => UriFactory::class,
+        'stream_factory' => StreamFactory::class,
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $useDiscovery = false;
+
+        foreach ($this->services as $service => $class) {
+            $serviceId = sprintf('httplug.%s.default', $service);
+
+            if (false === $container->has($serviceId)) {
+                // Register and create factory for the first time
+                if (false === $useDiscovery) {
```

Comment:
No, the `$factory` is a "httplug.factory". A service that works as a factory class for our *Factory classes. (MessageFactory, ClientFactory etc.. )